reference/test/multigrid/amgx_pgm_kernels.cpp

```diff
@@ -0,0 +1,595 @@
+              {I<T>({2.0, -1.0}), I<T>({-1.0, 2.0}), I<T>({0.0, -1.0}),
+               I<T>({3.0, -2.0}), I<T>({-2.0, 1.0})},
+              exec)),
+          coarse_b(gko::initialize<Vec>({I<T>({2.0, -1.0}), I<T>({0.0, -1.0})},
+                                        exec)),
+          restrict_ans((gko::initialize<Vec>(
+              {I<T>({0.0, -1.0}), I<T>({2.0, 0.0})}, exec))),
+          prolong_ans(gko::initialize<Vec>(
+              {I<T>({0.0, -2.0}), I<T>({1.0, -2.0}), I<T>({1.0, -2.0}),
+               I<T>({0.0, -1.0}), I<T>({2.0, 1.0})},
+              exec)),
+          fine_x(gko::initialize<Vec>(
+              {I<T>({-2.0, -1.0}), I<T>({1.0, -1.0}), I<T>({-1.0, -1.0}),
+               I<T>({0.0, 0.0}), I<T>({0.0, 2.0})},
+              exec)),
+          mtx(Mtx::create(exec, gko::dim<2>(5, 5), 15,
+                          std::make_shared<typename Mtx::classical>())),
+          weight(WeightMtx::create(
+              exec, gko::dim<2>(5, 5), 15,
+              std::make_shared<typename WeightMtx::classical>())),
+          coarse(Mtx::create(exec, gko::dim<2>(2, 2), 4,
+                             std::make_shared<typename Mtx::classical>())),
+          agg(exec, 5)
+    {
+        this->create_mtx(mtx.get(), weight.get(), &agg, coarse.get());
+        rstr_prlg = amgxpgm_factory->generate(mtx);
+        mtx_diag = weight->extract_diagonal();
+    }
+
+    void create_mtx(Mtx *fine, WeightMtx *weight, gko::Array<index_type> *agg,
+                    Mtx *coarse)
+    {
+        auto vals = fine->get_values();
+        auto cols = fine->get_col_idxs();
+        auto rows = fine->get_row_ptrs();
+        auto w_vals = weight->get_values();
+        auto w_cols = weight->get_col_idxs();
+        auto w_rows = weight->get_row_ptrs();
+        auto agg_val = agg->get_data();
+        auto c_vals = coarse->get_values();
+        auto c_cols = coarse->get_col_idxs();
+        auto c_rows = coarse->get_row_ptrs();
+        /* this matrix is stored:
+         *  5 -3 -3  0  0
+         * -3  5  0 -2 -1
+         * -3  0  5  0 -1
+         *  0 -3  0  5  0
+         *  0 -2 -2  0  5
+         */
+        vals[0] = 5;
+        vals[1] = -3;
+        vals[2] = -3;
+        vals[3] = -3;
+        vals[4] = 5;
+        vals[5] = -2;
+        vals[6] = -1;
+        vals[7] = -3;
+        vals[8] = 5;
+        vals[9] = -1;
+        vals[10] = -3;
+        vals[11] = 5;
+        vals[12] = -2;
+        vals[13] = -2;
+        vals[14] = 5;
+
+        rows[0] = 0;
+        rows[1] = 3;
+        rows[2] = 7;
+        rows[3] = 10;
+        rows[4] = 12;
+        rows[5] = 15;
+
+        cols[0] = 0;
+        cols[1] = 1;
+        cols[2] = 2;
+        cols[3] = 0;
+        cols[4] = 1;
+        cols[5] = 3;
+        cols[6] = 4;
+        cols[7] = 0;
+        cols[8] = 2;
+        cols[9] = 4;
+        cols[10] = 1;
+        cols[11] = 3;
+        cols[12] = 1;
+        cols[13] = 2;
+        cols[14] = 4;
```

Comment:
Maybe it is better to create a dense matrix and then convert it for easier readability ? You could also use the [initialize method](https://github.com/ginkgo-project/ginkgo/blob/develop/include/ginkgo/core/matrix/dense.hpp#L658) instead.

Reply:
Agree. Another way would be to create a CSR matrix from scratch and pass three arrays for the values, column indices and row pointer separately. Same thing for the next ones. At least that would save a lot of space.